storage/innobase/xtrabackup/src/xtrabackup.cc

```diff
@@ -5235,147 +5253,114 @@ xtrabackup_stats_func(int argc, char **argv)
 	/* gather stats */
 
 	{
-	dict_table_t*	sys_tables;
-	dict_index_t*	sys_index;
-	dict_table_t*	table;
-	btr_pcur_t	pcur;
-	rec_t*		rec;
-	byte*		field;
-	ulint		len;
-	mtr_t		mtr;
-
-	/* Enlarge the fatal semaphore wait timeout during the InnoDB table
-	monitor printout */
-
-	os_atomic_increment_ulint(
-		&srv_fatal_semaphore_wait_threshold,
-		72000);
-
-	mutex_enter(&(dict_sys->mutex));
-
-	mtr_start(&mtr);
-
-	sys_tables = dict_table_get_low("SYS_TABLES");
-	sys_index = UT_LIST_GET_FIRST(sys_tables->indexes);
-
-	btr_pcur_open_at_index_side(TRUE, sys_index, BTR_SEARCH_LEAF, &pcur,
-				    TRUE, 0, &mtr);
-loop:
-	btr_pcur_move_to_next_user_rec(&pcur, &mtr);
+		using Client = dd::cache::Dictionary_client;
+		using Releaser = dd::cache::Dictionary_client::Auto_releaser;
 
-	rec = btr_pcur_get_rec(&pcur);
+		my_thread_init();
+		THD *thd = create_thd(false,false,true,0);
 
-	if (!btr_pcur_is_on_user_rec(&pcur))
-	{
-		/* end of index */
+		dict_table_t*	sys_tables = nullptr;
+		dict_index_t*	sys_index = nullptr;
+		dict_table_t*	table = nullptr;
+		btr_pcur_t	pcur;
+		const rec_t*		rec = nullptr;
+		byte*		field;
+		ulint		len;
+		mtr_t		mtr;
+		MDL_ticket* mdl = nullptr;
+		mem_heap_t* heap = mem_heap_create(1000);
 
-		btr_pcur_close(&pcur);
-		mtr_commit(&mtr);
+		mutex_enter(&(dict_sys->mutex));
 
-		mutex_exit(&(dict_sys->mutex));
+		mtr_start(&mtr);
 
-		/* Restore the fatal semaphore wait timeout */
-		os_atomic_increment_ulint(
-			&srv_fatal_semaphore_wait_threshold,
-			-72000);
+		rec = dd_startscan_system(thd, &mdl, &pcur, &mtr, "mysql/tables", &sys_tables);
 
-		goto end;
-	}
+		while(rec) {
+			MDL_ticket* mdl_on_tab = nullptr;
+			dd_process_dd_tables_rec_and_mtr_commit(heap, rec, &table, sys_tables, &mdl_on_tab, &mtr);
+			mutex_exit(&(dict_sys->mutex));
+			if(table != nullptr)
+			{
+				// We have to skip some functions in the mysql db, 
```

Comment:
I have fixed this crash on mysql system tables with https://github.com/gl-sergei/percona-xtrabackup/commit/d72d2037b1803030aec1d4e4fd0ec712f60f4832